sheerwater_benchmarking/utils/caching.py

```diff
@@ -486,197 +644,75 @@ def cacheable(data_type, cache_args, timeseries=None, chunking=None, chunk_by_ar
-                elif backend == 'parquet':
-                    cache_path = "gs://sheerwater-datalake/caches/" + cache_key + '.parquet'
-                    null_path = "gs://sheerwater-datalake/caches/" + cache_key + '.null'
-                    supports_filepath = True
-                elif backend == 'postgres':
-                    cache_path = cache_key
-                    null_path = "gs://sheerwater-datalake/caches/" + cache_key + '.null'
-                    supports_filepath = False
-                else:
-                    raise ValueError("Only delta, parquet, and postgres backends are supported for tabular data")
-            elif data_type == 'basic':
-                backend = "pickle" if backend is None else backend
-                if storage_backend is None:
-                    storage_backend = backend
-                cache_path = "gs://sheerwater-datalake/caches/" + cache_key + '.pkl'
-                null_path = "gs://sheerwater-datalake/caches/" + cache_key + '.null'
-                supports_filepath = True
-            else:
-                raise ValueError("Caching currently only supports the 'array', 'tabular', and 'basic' datatypes")
 
-            if filepath_only and not supports_filepath:
-                raise ValueError(f"{backend} backend does not support filepath_only flag")
+    def __init__(self, data_type, cache_args,
+                 timeseries=None, validate_cache_timeseries=True,
+                 chunking=None, chunk_by_arg=None, auto_rechunk=False,
+                 cache=True, recompute=False, force_overwrite=False, retry_null_cache=False,
+                 cache_disable_if=None,
+                 filepath_only=False,
+                 backend=None, storage_backend=None):
+        """Initialize cacheable filesystems and configuration."""
+        # Initialize the google filesystem
+        self.fs = gcsfs.GCSFileSystem(project='sheerwater', token='google_default')
+
+        self.config = {
+            'data_type': data_type,
+            'cache_args': cache_args,
+            'timeseries': timeseries,
+            'validate_cache_timeseries': validate_cache_timeseries,
+            'chunking': chunking,
+            'chunk_by_arg': chunk_by_arg,
+            'auto_rechunk': auto_rechunk,
+            'cache': cache,
+            'recompute': recompute,
+            'force_overwrite': force_overwrite,
+            'retry_null_cache': retry_null_cache,
+            'cache_disable_if': cache_disable_if,
+            'filepath_only': filepath_only,
+            'backend': backend,
+            'storage_backend': storage_backend
+        }
+
+    def __call__(self, func):
+        """Decorate a function to be cacheable."""
+        @wraps(func)
+        def cacheable_wrapper(*args, **kwargs):
+            # The config that determines cacheable behavior, based on default and passed arguments
+            config = get_cache_config(self.config, func, args, kwargs)
```

Comment:
Why did this need to be abstracted like this?

I'm not opposed to abstraction, but I'm not fully seeing why this is where the line was drawn.

We could have for instance not had this abstraction and just referenced self.config for all state stored to this instance of cacheable right?